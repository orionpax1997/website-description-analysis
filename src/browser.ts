import chrome from 'chrome-aws-lambda';
import core from 'puppeteer-core';

const isDev = !process.env.AWS_REGION;
let _browser: core.Browser | null;

const exePath =
  process.platform === 'win32'
    ? 'C:\\Users\\Administrator\\scoop\\apps\\google-chrome\\current\\chrome.exe'
    : process.platform === 'linux'
    ? '/usr/bin/google-chrome'
    : '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';

interface Options {
  args: string[];
  executablePath: string;
  headless: boolean;
}

async function getOptions() {
  let options: Options;
  if (isDev) {
    options = {
      executablePath: exePath,
      headless: false,
      args: [],
    };
  } else {
    options = {
      args: chrome.args,
      executablePath: await chrome.executablePath,
      headless: chrome.headless,
    };
  }
  return options;
}

export async function getBrowser() {
  if (_browser) {
    return _browser;
  }
  const options = await getOptions();
  _browser = await core.launch(options);
  return _browser;
}
