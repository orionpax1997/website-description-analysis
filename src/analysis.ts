import request from 'request';
import * as cheerio from 'cheerio';
import { CheerioAPI } from 'cheerio';
import iconv from 'iconv-lite';
import { getBrowser } from './browser';

/**
 * 解析工厂类
 * @author OrionPax
 */
export class AnalysisFactory {
  /**
   * 解析实例创建方法
   * @param url 要解析的网页地址
   * @returns 返回负责解析的实例
   * @author OrionPax
   */
  static async create(url: string): Promise<Analysis> {
    if (url.indexOf('juejin') !== -1) return await new JueJinAnalysisImpl(url).init();
    if (url.indexOf('segmentfault') !== -1) return await new SegmentfaultAnalysisImpl(url).init();
    return await new AnalysisImpl(url).init();
  }
}

/**
 * 解析类
 * @author OrionPax
 */
export abstract class Analysis {
  /** 目标 URL */
  protected readonly _url: string;
  /** 目标 HTML */
  protected _html!: string;
  /** CheerioAPI 对象 */
  protected $!: CheerioAPI;

  constructor(url: string) {
    this._url = url;
  }

  /**
   * 初始化方法
   * @returns 返回解析类本身
   * @author OrionPax
   */
  async init(): Promise<Analysis> {
    this._html = await curl(this._url);
    if (/<meta.*(gbk)+.*>/gi.test(this._html) && this._html.indexOf('�') !== -1) {
      this._html = await curl(this._url, 'gbk');
    }
    this.$ = cheerio.load(this._html);
    return this;
  }

  /**
   * 解析方法
   * @returns 返回解析数据
   * @see AnalysisData
   * @author OrionPax
   */
  async analysis(): Promise<AnalysisData> {
    return {
      url: this.url,
      title: this.title,
      description: this.description,
      image: this.image,
      favicon: await getBase64(this.favicon),
    };
  }

  /**
   * get URL
   * @author OrionPax
   */
  get url(): string {
    return this._url;
  }

  /**
   * 解析 Title
   * @author OrionPax
   */
  get title(): string {
    let article;
    if (this.$('article').length > 0) {
      article = this.$('article').first();
    } else if (this.$('div[class*=content]').length > 0) {
      article = this.$('div[class*=content]').first();
    } else if (this.$('div[class*=Content]').length > 0) {
      article = this.$('div[class*=content]').first();
    }
    if (this.$('meta[property=og:title]').length === 1) {
      return this.$('meta[property=og:title]').attr('content') as string;
    }
    if (article && article.find('h1').length === 1) {
      return article.find('h1').text().trim();
    }
    if (this.$('h1[class*=title]').length === 1) return this.$('h1[class*=title]').text().trim();
    if (this.$('h1[class*=Title]').length === 1) return this.$('h1[class*=Title]').text().trim();
    if (this.$('.title').length === 1) return this.$('.title').text().trim();
    if (this.$('h1>a').length === 1) return this.$('h1>a').text().trim();
    if (this.$('header h1').length === 1) return this.$('header h1').text().trim();
    if (this.$('h1').length === 1) return this.$('h1').text().trim();
    return this.$('title').text().trim();
  }

  /**
   * 解析 Description
   * @author OrionPax
   */
  get description(): string | undefined {
    let article;
    if (this.$('article').length > 0) {
      article = this.$('article').first();
    } else if (this.$('div[class*=content]').length > 0) {
      article = this.$('div[class*=content]').first();
    } else if (this.$('div[class*=Content]').length > 0) {
      article = this.$('div[class*=content]').first();
    }
    if (this.$('meta[property=og:description]').length === 1)
      return this.$('meta[property=og:description]').attr('content');
    else if (this.$('meta[name=description]').length === 1) {
      return this.$('meta[name=description]').attr('content');
    } else if (article && article.find('p').length > 0) {
      const p = article
        .find('p')
        .toArray()
        .find(p => {
          this.$(p).text().trim().length > 120;
        });
      if (p) {
        return this.$(p).text().trim().substring(0, 240);
      }
    }
    if (this.$('body p').length > 0) {
      const p = this.$('body p')
        .toArray()
        .find(p => this.$(p).text().trim().length > 120);
      if (p) {
        return this.$(p).text().trim().substring(0, 240);
      }
    }
    return this.$('body').text().substring(0, 240);
  }

  /**
   * 解析分享图或文章首张图片
   * @author OrionPax
   */
  get image(): string | undefined {
    let img;
    let article;
    if (this.$('article').length > 0) {
      article = this.$('article').first();
    } else if (this.$('div[class*=content]').length > 0) {
      article = this.$('div[class*=content]').first();
    } else if (this.$('div[class*=Content]').length > 0) {
      article = this.$('div[class*=content]').first();
    }
    if (this.$('meta[name=og:image]').length === 1) {
      img = this.$('meta[name=og:image]').attr('content');
    } else if (this.$('meta[property=og:image]').length === 1) {
      img = this.$('meta[property=og:image]').attr('content');
    } else if (article && article.find('p img').length > 0) {
      img = article.find('p img').first().attr('src');
    } else if (article && article.find('img').length > 0) {
      img = article.find('img').first().attr('src');
    }

    if (!img && this.$('img').length > 0) {
      const item = this.$('img')
        .toArray()
        .find(item => this.$(item).parent()[0].tagName === 'p');
      if (item) {
        img = this.$(item).attr('src');
      }
    }
    if (!img && this.$('img').length > 0) {
      const item = this.$('img')
        .toArray()
        .find(
          item => this.$(item).attr('src')?.indexOf('banner') !== -1 || this.$(item).attr('src')?.indexOf('logo') !== -1
        );
      if (item) {
        img = this.$(item).attr('src');
      }
    }

    // 相对路径转绝对路径
    if (img && !img.startsWith('http')) {
      img = resolve(this._url, img);
    }
    return img;
  }

  /**
   * 解析 Favicon
   * @author OrionPax
   */
  get favicon(): string {
    return `https://t3.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&fallback_opts=TYPE,SIZE,URL&url=${this._url}&size=16`;
  }
}

/**
 * 默认实现类
 */
class AnalysisImpl extends Analysis {}

/**
 * 掘金的实现类
 * @see https://juejin.cn/
 * @author OrionPax
 */
class JueJinAnalysisImpl extends Analysis {
  private readonly titleRegExp: RegExp = /(?<=headline":")[^"]*/;
  private readonly imageRegExp: RegExp = /(?<=src=\\")[^"]*/;

  /**
   * 获取 Title
   * @author OrionPax
   */
  get title(): string {
    const match = this.titleRegExp.exec(this._html);
    if (match) return match[0];
    return '';
  }

  /**
   * @author OrionPax
   * 获取分享图或文章首张图片
   */
  get image(): string | undefined {
    const match = this.imageRegExp.exec(this._html);
    if (match) return unescape(match[0].replace(/\\u/g, '%u')).replace('\\', '');
    return undefined;
  }
}

/**
 * Segmentfault 的实现类
 * @see https://segmentfault.com
 * @author OrionPax
 */
class SegmentfaultAnalysisImpl extends Analysis {
  /**
   * 初始化方法
   * @returns 返回解析类本身
   * @author OrionPax
   */
  async init(): Promise<Analysis> {
    this._html = await getHtml(this._url, '.container');
    this.$ = cheerio.load(this._html);
    return this;
  }

  get description(): string | undefined {
    return this.$('meta[name=description]').attr('content');
  }

  get image(): string | undefined {
    return undefined;
  }
}

/**
 * 解析数据对象
 * @see Analysis.analysis
 * @author OrionPax
 */
export type AnalysisData = {
  /** 网站 URL */
  readonly url: string;
  /** 网站标题 */
  readonly title: string;
  /** 网站描述信息 */
  readonly description: string | undefined;
  /** 网站分享图或正文第一张图片 */
  readonly image: string | undefined;
  /** 网站 Favicon */
  readonly favicon: string | undefined;
};

/**
 * 获取目标 URL 的 HTML 代码
 * @param url 目标 URL
 * @returns 目标 HTML
 * @author OrionPax
 */
async function curl(url: string, encoding?: string): Promise<string> {
  return new Promise((resolve, reject) =>
    request(
      {
        url,
        headers: {
          Accept:
            'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
          'user-agent':
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36',
          Connection: 'keep-alive',
        },
        encoding: encoding !== undefined ? null : 'utf-8',
      },
      async (error, res, body) => {
        if (error) {
          reject(error);
        }
        if (res.statusCode === 301) {
          resolve(await curl(res.headers.location as string, encoding));
        }
        if (res.statusCode !== 200) {
          reject(res.statusMessage);
        }
        if (encoding) {
          resolve(iconv.decode(body, encoding));
        } else {
          resolve(body);
        }
      }
    )
  );
}

async function getHtml(url: string, selector: string) {
  const browser = await getBrowser();
  const page = await browser.newPage();
  await page.goto(url);
  await page.waitForSelector(selector, { visible: true });
  const pageContent = await page.content();
  await page.close();
  return pageContent;
}

/**
 * 相对路径转绝对路径
 * @author OrionPax
 */
function resolve(url: string, relative: string): string {
  return new URL(relative, url).toString();
}

/**
 * 解析图片为 base64
 * @param url 图片 URL
 * @returns Base64 编码的图片
 * @author OrionPax
 */
async function getBase64(url: string): Promise<string | undefined> {
  return new Promise(resolve =>
    request({ url, encoding: null }, (error, response, body) => {
      if (error || response.statusCode !== 200) {
        resolve(undefined);
      } else {
        resolve(`data:${response.headers['content-type']};base64,${Buffer.from(body).toString('base64')}`);
      }
    })
  );
}
