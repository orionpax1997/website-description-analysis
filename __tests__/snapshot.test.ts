import { AnalysisFactory } from '../src/analysis';
import { getBrowser } from '../src/browser';

describe('get description function', () => {
  it('analysis https://segmentfault.com/a/1190000013568216', async () => {
    const analysisData = await (await AnalysisFactory.create('https://segmentfault.com/a/1190000013568216')).analysis();
    expect(analysisData).toMatchSnapshot();
  });

  it('analysis https://blog.csdn.net/design_Lu/article/details/94870265', async () => {
    const analysisData = await (
      await AnalysisFactory.create('https://blog.csdn.net/design_Lu/article/details/94870265')
    ).analysis();
    expect(analysisData).toMatchSnapshot();
  });

  it('analysis https://www.jianshu.com/p/eec5e34ff0c2', async () => {
    const analysisData = await (await AnalysisFactory.create('https://www.jianshu.com/p/eec5e34ff0c2')).analysis();
    expect(analysisData).toMatchSnapshot();
  });

  it('analysis https://blog.51cto.com/u_15349616/3717558', async () => {
    const analysisData = await (await AnalysisFactory.create('https://blog.51cto.com/u_15349616/3717558')).analysis();
    expect(analysisData).toMatchSnapshot();
  });

  it('analysis https://juejin.cn/post/6844904009887645709', async () => {
    const analysisData = await (await AnalysisFactory.create('https://juejin.cn/post/6844904009887645709')).analysis();
    expect(analysisData).toMatchSnapshot();
  });

  it('analysis http://t.mb5u.com/html/', async () => {
    const analysisData = await (await AnalysisFactory.create('http://t.mb5u.com/html/')).analysis();
    expect(analysisData).toMatchSnapshot();
  });

  afterAll(async () => {
    const browser = await getBrowser();
    if (browser) {
      browser.close();
    }
  });
});
