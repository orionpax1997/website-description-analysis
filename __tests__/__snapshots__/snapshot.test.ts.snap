// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get description function analysis http://t.mb5u.com/html/ 1`] = `
{
  "description": "html参考手册,包括HTML4和HTML5手册。",
  "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABYUlEQVQ4jaWSQU7CUBCGv7YPK9IClaCwcGM8gntjVx5Aly5cGW7gTRpXHMALsNK4NPEALjTRmIBKYqAoBOh7zwXYtGlZGGc1b/6Z//9n8ozdi/Yx0AJ8/hbXQCCA1v7hgb+5vfWn6c/3D//+5hYB+BOnxtMwSjX0R5JwogAoF03qrpXC150agC8AxjOVURh+Sx5P6gDsXfWplcwUPp5pAARAlJ1HyXSe1xMTSKUzgJY6lef1JBxoBmHE1zhfRinFc3cav50Nk2pZJAgkhKHi9ayRS/Bymq7vtN9wSovc/HWgZL7FXEdSE6nEEaWCgjBoXvZSjb3zJkCmbtsGcrmtAFBa41Qt7KlCLc89GiWOqDWuaywsC5OCbaK0ThIApkGhaAGLDxOGM3RMALZbSK+hEzf4T4hVwJpt0Ai6cb4qTKCjBv0M4HqCimdR8SxcL6uznOkIIJg/3AEcrVKZ55c7QPADejmLwEGbWDEAAAAASUVORK5CYII=",
  "image": undefined,
  "title": "HTML参考手册",
  "url": "http://t.mb5u.com/html/",
}
`;

exports[`get description function analysis https://blog.51cto.com/u_15349616/3717558 1`] = `
{
  "description": "",
  "image": undefined,
  "title": "",
  "url": "https://blog.51cto.com/u_15349616/3717558",
}
`;

exports[`get description function analysis https://blog.csdn.net/design_Lu/article/details/94870265 1`] = `
{
  "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAAXVBMVEX8VTH////8SBv8RRT8Ui38UCn+4dr8PQD8RBL9lYL/8u/+5N78Sh78f2b+0cn+7Of9xbv8Xzv8Qgr8blH9uaz+2NH8hW38ZEP8TiX9vbP/8e3+6eP8WDT8aUr9jHdkgCI0AAAAgElEQVQYlW2O2xKDIAxEWQKSKBXqpSK9/P9n1o7Uykzztie7ySr1f6xm8uan6dLnEK/2q4cINC0wFo+eMGtygttuMQuy23YplYhfIen8gRqYs1ac4SpAd+gK+Aj5EJseOzBP4OXYjQFcLAJg64WOSigta9eGXobjjPHEpKsyx7wBKpEEM9DQtpwAAAAASUVORK5CYII=",
  "image": "https://i-blog.csdnimg.cn/blog_migrate/83d4518b12eb78d3f7c9785391201c8e.png",
  "title": "vue生命周期和react生命周期对比。",
  "url": "https://blog.csdn.net/design_Lu/article/details/94870265",
}
`;

exports[`get description function analysis https://juejin.cn/post/6844904009887645709 1`] = `
{
  "description": "主要从githubstarts&issues量，npm下载量，使用方式，以及语法示例对三款测试框架做了一下较简单的比较Githubstars&issuesNpm下载量使用方式框架断言异步Mock代码覆",
  "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABVElEQVQ4jaXTPUgbUBAH8F9MoIJgdCiddLAiDn7MajcVJ7fOurgooovfoKSllZa6OAmisw5CFRGqglPAVUTEqZTSqcXBDsVF4pALxqTq4IPjHffu/u9/XzzzJP5nrM/kCmpD3N/hZ6bcveKR4B58DekueXsYoIhZGxpD2h9imyhBfokWHOMGb8O+hSQ6cIY/hZSKGSQxiV18RDU2QtJYjLeJ8M0zqM/kkniBf2jFJ/ThWwDCUtj2MR0sqnCdwih6IzCLQcxhGNsBUIflYHGJN5jFQQVq0BVUB3CFGYwH9erQZ/E3PthEJ9IpfMAJvmAFzaGv4jwYZFGLKYzhF0awl4pq7+BH5DoVINMRCE34jH4cRSFPKW/jK8xjCBfhWChiM9bwHr/Jt/HecARQZQAsuGvXDd5hHdfFI102XQGSkO/MUpgncIhc6T48tUyvkfPIMj373AJGC1eDWBe94QAAAABJRU5ErkJggg==",
  "image": "https://p1-jj.byteimg.com/tos-cn-i-t2oaga2asx/gold-user-assets/2019/12/3/16eca535239e77d7~tplv-t2oaga2asx-jj-mark:3024:0:0:0:q75.png",
  "title": "JS测试框架Jest/Mocha/Ava的简单比较",
  "url": "https://juejin.cn/post/6844904009887645709",
}
`;

exports[`get description function analysis https://segmentfault.com/a/1190000013568216 1`] = `
{
  "description": "Hibernate是一种ORM框架，全称为 Object_Relative DateBase-Mapping，在Java对象与关系数据库之间建立某种映射，以实现直接存取Java对象！",
  "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAALVBMVEUAmmEAmmAAlFX///8Aj0s+roHI6Nuu2cYhpXL1/ftZtY7p+POc1r+Ax6pvu5gxFaaeAAAAAXRSTlO7HFWyYgAAAGxJREFUGJVVj0sSgDAIQ0vSQj/q/Y8rYkdbFjB587IgJdkmLRk+Tv58kgMLEFKPFRQyzwpKEQSQALWRPKov2lOBUrv1ak4vmQAu16/iZWocvADituIHBbm7/gG0ZmMzurZhLqpOML+SWNu7kW9dnQK2J4kwSgAAAABJRU5ErkJggg==",
  "image": undefined,
  "title": "Hibernate入门这一篇就够了",
  "url": "https://segmentfault.com/a/1190000013568216",
}
`;

exports[`get description function analysis https://www.jianshu.com/p/eec5e34ff0c2 1`] = `
{
  "description": "在 TypeScript 开发中，我们需要为开发的模块编写单元测试，可以使用 jest 这个框架。可参考 npm - jest 如果你已经有一个 TypeScript 项目，...",
  "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAAOVBMVEXlYEj76OXiQBj99PL////lYEjtlIXnYkjkW0HkUjXsiXn41M763tnmX0boc17gLwDyt67tmIrobVcZOABUAAAACHRSTlPq////////MiIP05YAAACBSURBVBiVXY9bEsMgDAMxVojFu7n/YYtpp5Nmv9COwSic4bgRzrCysa0zuY3njEZSanS3BFUgklAhlUsQc9QZ0QfynmjX1KyOSDa/cjRNV3eQXNgLo2BR41ckxFrGKHITBar4E8+J9QVBjr9HuekJe63lslvYlMldxvhpajTPz/pvg+IGkoY/qWQAAAAASUVORK5CYII=",
  "image": "https://upload-images.jianshu.io/upload_images/1362764-9e4cc8a17ba4048b.png",
  "title": "TypeScript + jest + VSCode 编写单元测试",
  "url": "https://www.jianshu.com/p/eec5e34ff0c2",
}
`;

exports[`get description function analysis https://www.jianshu.com/p/eec5e34ff0c2 2`] = `
{
  "description": "在 TypeScript 开发中，我们需要为开发的模块编写单元测试，可以使用 jest 这个框架。可参考 npm - jest 如果你已经有一个 TypeScript 项目，...",
  "favicon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAAOVBMVEXlYEj76OXiQBj99PL////lYEjtlIXnYkjkW0HkUjXsiXn41M763tnmX0boc17gLwDyt67tmIrobVcZOABUAAAACHRSTlPq////////MiIP05YAAACBSURBVBiVXY9bEsMgDAMxVojFu7n/YYtpp5Nmv9COwSic4bgRzrCysa0zuY3njEZSanS3BFUgklAhlUsQc9QZ0QfynmjX1KyOSDa/cjRNV3eQXNgLo2BR41ckxFrGKHITBar4E8+J9QVBjr9HuekJe63lslvYlMldxvhpajTPz/pvg+IGkoY/qWQAAAAASUVORK5CYII=",
  "image": "https://upload-images.jianshu.io/upload_images/1362764-9e4cc8a17ba4048b.png",
  "title": "TypeScript + jest + VSCode 编写单元测试",
  "url": "https://www.jianshu.com/p/eec5e34ff0c2",
}
`;
